[geoclue-provider-mlsdb] Use location code in lookups. Contributes to JB#34744

chriadam opened the pull request:
Previously, the plugin assumed that the cellId was globally unique.
This is not true, instead the cellId is only unique within a given
location code (LAC on GSM or TAC on LTE networks).

This commit ensures that the key for location lookups is composed
of the location area code and the cell id.

It also regenerates the databases so that they contain the appropriate
data.

Contributes to JB#34744

saukko commented on plugin/mlsdbprovider.cpp:
```diff
@@ -269,21 +269,29 @@ void MlsdbProvider::calculatePositionAndEmitLocation()
+            cellId = static_cast<quint32>(c->ci());
         } else {
-            qCDebug(lcGeoclueMlsdbPosition) << "ignoring neighbour cell with no cell id:" << c->type() << c->mcc() << c->mnc();
+            qCDebug(lcGeoclueMlsdbPosition) << "ignoring neighbour cell with no cell id:" << c->type() << c->mcc() << c->mnc() << c->lac() << c->tac();
```
Would be good to add here also what the values are to make easier to debug. something like 

qCDebug(lcGeoclueMlsdbPosition) << "ignoring neighbour cell with no cell id. type:" << c->type() << ", mcc: " << c->mcc() << ", mnc: " << c->mnc();

chriadam replied:
will do

chriadam commented:
@monich @saukko @pvuorela

saukko commented on plugin/mlsdbprovider.cpp:
```diff
@@ -269,21 +269,29 @@ void MlsdbProvider::calculatePositionAndEmitLocation()
-            qCDebug(lcGeoclueMlsdbPosition) << "have neighbour cell:" << cell.cellId << "with strength:" << c->signalStrength();
+        cell.composedCellId = composeMlsdbCellId(cellType, locationCode, cellId);
+        if (!seenCellIds.contains(cell.composedCellId)) {
+            qCDebug(lcGeoclueMlsdbPosition) << "have neighbour cell:" << stringForMlsdbCellType(cellType) << locationCode << cellId << "with strength:" << c->signalStrength();
```
Also here maybe state the mcc, type etc. info so one can easily track it? 

qCDebug(lcGeoclueMlsdbPosition) << "have neighbour cell with type:" << c->type() << ", cell id: " << cell.cellId << ", mcc: " << c->mcc() << ", mnc: " << c->mnc() << "with strength:" << c->signalStrength();

saukko commented:
Tested and now I get right location instantly \o/ This :+1: is mainly for functionality not for code itself

pvuorela commented on plugin/mlsdbprovider.cpp:
```diff
@@ -269,78 +272,106 @@ void MlsdbProvider::calculatePositionAndEmitLocation()
         }
+        // we have a known location for this cell.  Update our locations list.
+        cellLocations.insert(cell.composedCellId, cellCoords);
+        totalSignalStrength += (1.0 * cell.signalStrength);
```
Existing code, but wonder why this 1.0.

chriadam replied:
to make it explicitly a double value

pvuorela commented:
LGTM

monich commented on plugin/mlsdbprovider.cpp:
```diff
@@ -269,78 +272,106 @@ void MlsdbProvider::calculatePositionAndEmitLocation()
         if (c->cid() != -1 && c->cid() != 0) {
             // gsm / wcdma
-            cell.cellId = c->cid();
+            cellType = MLSDB_CELL_TYPE_GSM_WCDMA;
```
Why guess the cell type this way if it's available from `c->type()`?

chriadam replied:
I have to check the value of cid() / ci() anyway, since we often get invalid values like "ignoring neighbour cell with no cell id: 3 -1 -1 -1 -1" reported (i.e., that one's an LTE cell with invalid values for the various fields), so implicitly detecting the type doesn't cost us anything but saves us one needless check.

chriadam commented:
Updated to include the mcc+mnc in the cell id key, and also now supports UMTS cells properly (as well as LTE+GSM like before).  Updated the data files.

Should be ready for review.

mlehtima commented:
Tested on Fairphone 2 and after the latest changes cell type changes it works again in my location.

chriadam commented:
Rebased.  Anyone have any issues with me merging this one?

chriadam commented:
Great, merging now.
